include/nbl/asset/ICPUGraphicsPipeline.h

```diff
@@ -65,39 +103,45 @@ class ICPUGraphicsPipeline final : public ICPUPipeline<IGraphicsPipeline<ICPUPip
 		using base_t::base_t;
         ~ICPUGraphicsPipeline() = default;
 
-		base_t* clone_impl(core::smart_refctd_ptr<const ICPUPipelineLayout>&& layout) const override
-		{
-			std::array<IPipelineBase::SShaderSpecInfo,GRAPHICS_SHADER_STAGE_COUNT> _shaders;
-			for (auto i=0; i<GRAPHICS_SHADER_STAGE_COUNT; i++)
-				_shaders[i] = m_stages[i].info;
-			const SCreationParams params = {{
-				.shaders = _shaders,
-				.cached = m_params,
-				.renderpass = m_renderpass.get()
-			}};
-			return new ICPUGraphicsPipeline(params);
-		}
 		inline IAsset* getDependant_impl(const size_t ix) override
 		{
 			if (ix==0)
 				return const_cast<ICPUPipelineLayout*>(m_layout.get());
 			if (ix==1)
 				return m_renderpass.get();
 			size_t stageCount = 0;
-			for (auto& stage : m_stages)
-			if (stage.shader)
-			if ((stageCount++)==ix-2)
-				return stage.shader.get();
+			for (auto& specInfo : m_specInfos)
+			{
+        if (specInfo.shader)
+        {
+          if ((stageCount++)==ix-2)
+            return specInfo.shader.get();
+        }
+			}
 			return nullptr;
 		}
 
-		inline int8_t stageToIndex(const hlsl::ShaderStage stage) const override
+		inline int8_t stageToIndex(const hlsl::ShaderStage stage) const
 		{
 			const auto stageIx = hlsl::findLSB(stage);
-			if (stageIx<0 || stageIx>=GRAPHICS_SHADER_STAGE_COUNT || hlsl::bitCount(stage)!=1)
+			if (stageIx<0 || stageIx>= GRAPHICS_SHADER_STAGE_COUNT || hlsl::bitCount(stage)!=1)
 				return -1;
 			return stageIx;
 		}
+
+    inline bool setSpecInfo(const SShaderSpecInfo<true>& info)
+		{
+			assert(isMutable());
+      const auto specSize = info.valid();
+      if (specSize<0) return false;
+			const auto stage = info.stage;
+			const auto stageIx = stageToIndex(stage);
+			if (stageIx<0) return false;
+			m_specInfos[stageIx] = info;
+			return true;
+		}
```

Comment:
I'd change the design a bit, and allow any stage in the spec info, and we check validity later

Reply:
overruled by discord conversation